modules/social_features/social_event/src/EdaHandler.php

```diff
@@ -0,0 +1,99 @@
+   * {@inheritDoc}
+   */
+  public function __construct(
+    private ?SocialEdaDispatcher $dispatcher,
+    private ?SocialEdaDataFormatter $formatter,
+    private UuidInterface $uuid,
+    private RequestStack $requestStack,
+  ) {}
+
+  /**
+   * Create event handler.
+   */
+  public function eventCreate(Event $event): void {
+    if ($this->dispatcher && $this->formatter) {
+      // Get current request.
+      $request = $this->requestStack->getCurrentRequest();
+
+      // List enrollment methods.
+      $enrollment_methods = ['open', 'request', 'invite'];
+
+      // The event data array.
+      $event_data = [
+        'id' => $event->uuid->value,
+        'created' => $this->formatter->getFormattedDrupalDateTime($event->created->value),
+        'updated' => $this->formatter->getFormattedDrupalDateTime($event->changed->value),
+        'status' => $event->status->value,
+        'label' => (string) $event->label(),
+        'visibility' => $event->field_content_visibility->value,
+        'group' => !$event->groups->isEmpty() ? $this->formatter->getFormattedEntity($event->groups->entity) : NULL,
+        'author' => $this->formatter->getFormattedUser($event->uid->entity),
+        'allDay' => $event->field_event_all_day->value,
+        'start' => $event->field_event_date->value,
+        'end' => $event->field_event_date_end->value,
+        'timezone' => date_default_timezone_get(),
+        'contentTags' => [],
+        'address' => [
+          'label' => $event->field_event_location->value,
+          ...$this->formatter->getFormattedAddress($event->field_event_location->getValue()),
+        ],
+        'enrollment' => [
+          'enabled' => (bool) $event->field_event_enroll->value,
+          'method' => $enrollment_methods[$event->field_enroll_method->value],
+        ],
+        'href' => $this->formatter->getFormattedEntityHref($event),
+        'type' => $event->hasField('field_event_type') && !$event->field_event_type->isEmpty() ? $event->field_event_type->entity->label() : NULL,
+      ];
+
+      // Populate `contentTags`.
+      if (!$event->social_tagging->isEmpty()) {
+        foreach ($event->social_tagging->referencedEntities() as $tag) {
+          $event_data['contentTags'][] = $tag->label();
+        }
+      }
+
+      // The complete data object.
+      $data = new CloudEvent(
+        $this->uuid->generate(),
+        $request->getUri(),
+        'com.getopensocial.cms.event.create',
+        [
+          'event' => new EventCreateEventData(...$event_data),
+          'actor' => [
+            'application' => NULL,
+            'user' => $this->formatter->getFormattedUser($event->uid->entity),
+          ],
+        ],
+        'application/json',
+        NULL,
+        NULL,
+        $this->formatter->getFormattedDateTimeImmutable($event->created->value),
+
+      );
+
+      // Dispatch the event to the message broker.
+      $this->dispatcher->dispatch('com.getopensocial.cms.event.create', $data);
```

Comment:
We can re-arrange the content tagging a bit to put it at the top in an intermediary variable. This allows us to inline all other variables.

Rather than using an array and array spread, we can use named arguments from the underlying class. This has the benefit of providing you with suggestions when you type it out and will highlight properties that are misspelled or don't exist on the class.

```suggestion
      // If content tagging is enabled they're added to the event.
      $contentTags = [];
      if ($event->hasField('social_tagging') && !$event->social_tagging->isEmpty()) {
        foreach ($event->social_tagging->referencedEntities() as $tag) {
          $contentTags[] = $tag->label();
        }
      }

      $this->dispatcher->dispatch(
        'com.getopensocial.cms.event.create',
        new CloudEvent(
          $this->uuid->generate(),
          $request->getUri(),
          'com.getopensocial.cms.event.create',
          [
            'event' => new EventCreateEventData(
              id: $event->uuid->value,
              created: $this->formatter->getFormattedDrupalDateTime($event->created->value),
              updated: $this->formatter->getFormattedDrupalDateTime($event->changed->value),
              status: $event->status->value,
              label: (string) $event->label(),
              visibility: $event->field_content_visibility->value,
              group: !$event->groups->isEmpty() ? $this->formatter->getFormattedEntity($event->groups->entity) : NULL,
              author: $this->formatter->getFormattedUser($event->uid->entity),
              allDay: $event->field_event_all_day->value,
              start: $event->field_event_date->value,
              end: $event->field_event_date_end->value,
              timezone: date_default_timezone_get(),
              contentTags: $contentTags,
              address: [
                'label' => $event->field_event_location->value,
                ...$this->formatter->getFormattedAddress($event->field_event_location->getValue()),
              ],
              enrollment: [
                'enabled' => (bool) $event->field_event_enroll->value,
                'method' => $enrollment_methods[$event->field_enroll_method->value],
              ],
              href:  $this->formatter->getFormattedEntityHref($event),
              type: $event->hasField('field_event_type') && !$event->field_event_type->isEmpty() ? $event->field_event_type->entity->label() : NULL,
            ),
            'actor' => [
              'application' => NULL,
              'user' => $this->formatter->getFormattedUser($event->uid->entity),
            ],
          ],
          'application/json',
          NULL,
          NULL,
          $this->formatter->getFormattedDateTimeImmutable($event->created->value),
        )
      );
```

Additionally, I'd go a step further and also create classes for the things that are now structured arrays. For exmaple `EventCreateAuthorEventData` (or `AuthorEventData` if we're 100% sure this is a standardised piece of data).

That also allows you to eliminate the formatter because you can make create a static method `AuthorEventData::fromUser($user_entity)` which contains what is now in `DataFormatter::getFormattedUser`. The DataFormatter service can likely be entirely eliminated in this way.

The benefit is that your entire tree of data becomes statically defined and analysable by PHPStan. If you need to access the data anywhere then it's also consistently accessable with the object method (`$event->data->author->name` rather than having something like `$event->data->author['name']`).

Reply:
Done. Thanks :)